src/saltext/vmware/utils/license_mgr.py

```diff
@@ -0,0 +1,448 @@
+        return False
+
+    return None
+
+
+def get_license_mgr(service_instance):
+    """
+    Return a license manager from specified Service Instance
+
+    service_instance
+        vCenter service connection instance
+
+    Note: returns LicenseManager
+    """
+    lic_mgr = None
+
+    if not is_vcenter(service_instance):
+        return lic_mgr
+
+    srv_content = get_service_content(service_instance)
+    lic_mgr = srv_content.licenseManager
+    return lic_mgr
+
+
+def get_license_assignment_mgr(service_instance):
+    """
+    Return a license assignment manager from specified Service Instance
+
+    service_instance
+        vCenter service connection instance
+
+    Note: returns LicenseAssignmentManager
+    """
+    lic_mgr = None
+
+    if not is_vcenter(service_instance):
+        return lic_mgr
+
+    srv_content = get_service_content(service_instance)
+    lic_mgr = srv_content.licenseManager.licenseAssignmentManager
+    return lic_mgr
+
+
+def list_licenses(service_instance):
+    """
+    Returns a list of licenses associated with specified Service Instance.
+
+    service_instance
+        The Service Instance Object from which to obtain licenses.
+    """
+    log.debug("start list of License Managers")
+
+    ret = {}
+    lic_mgr = get_license_mgr(service_instance)
+    log.debug(f"License Manager listing of licenses '{lic_mgr.licenses}'")
+
+    if not lic_mgr:
+        ret["message"] = "Failed, not connected to a vCenter"
+        ret["result"] = False
+        return ret
+
+    ret["licenses"] = _list_lic_keys(lic_mgr.licenses)
+    return ret
+
+
+def add_license(
+    service_instance, license_key, datacenter_name=None, cluster_name=None, esxi_hostname=None
+):
+    """
+    Add license to the pool of available licenses associated with the License Manager
+    and assign it to the specified Cluster, ESXI Server or vCenter
+
+    If no datacenter, cluster or ESXI Server is specified, it is assumed the operation is to be applied to a vCenter
+
+    service_instance:
+        Service Instance containing a License Manager
+
+    license_key
+        License key which specifies license to add to license manager
+
+    datacenter_name
+        Datacenter name to use for the operation [default None]
+
+    cluster_name
+        Name of the cluster to add license [default None]
+
+    esxi_hostname
+        Hostname of the ESXI Server to add license [default None]
+
+    Returns:
+        True - if successful or license already present
+    """
+    lic_mgr = get_license_mgr(service_instance)
+    if not lic_mgr:
+        return False
+
+    lic_keys = _list_lic_keys(lic_mgr.licenses)
+    log.debug(
+        f"attempting to add license key '{license_key}' to list of existing license keys '{lic_keys}'"
+    )
+
+    try:
+        if not license_key in lic_keys:
+            licMgrLicInfo = lic_mgr.AddLicense(licenseKey=license_key)
+            log.debug(f"attempted AddLicense, returned license into '{licMgrLicInfo}'")
```

Comment:
that is just it, the call doesn't error out or even throw exceptions which I would have expected it should, they have throw Fault in their doc., but nope returns LicenseInfo which you then have to examine.

Had a contact on their team check all this since I was flabbergasted, could add license_key 'mydog' and it would not throw anything, where I was then directed to LicenseInfo returned. so 'attempted' since it is past-tense, and seeing result of attempt.

Reply:
Well. That is.... interesting :joy: 

I'd say it sounds good, but... it sounds like you're doing the correct thing here then :upside_down_face: